src/main/kotlin/io/github/gradlenexus/publishplugin/InitializeNexusStagingRepository.kt

```diff
@@ -16,67 +16,23 @@
 
 package io.github.gradlenexus.publishplugin
 
-import io.github.gradlenexus.publishplugin.internal.NexusClient
 import io.codearte.gradle.nexus.NexusStagingExtension
-import org.gradle.api.DefaultTask
-import org.gradle.api.GradleException
+import io.github.gradlenexus.publishplugin.internal.NexusClient
 import org.gradle.api.artifacts.repositories.MavenArtifactRepository
 import org.gradle.api.model.ObjectFactory
 import org.gradle.api.provider.Property
 import org.gradle.api.publish.PublishingExtension
-import org.gradle.api.tasks.Input
-import org.gradle.api.tasks.Internal
-import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.TaskAction
-import org.gradle.kotlin.dsl.property
 import org.gradle.kotlin.dsl.the
 import java.net.URI
-import java.time.Duration
 import javax.inject.Inject
 
 @Suppress("UnstableApiUsage")
 open class InitializeNexusStagingRepository @Inject
-constructor(objects: ObjectFactory, extension: NexusPublishExtension, repository: NexusRepository, private val serverUrlToStagingRepoUrl: MutableMap<URI, URI>) : DefaultTask() {
-
-    @get:Input
-    val serverUrl: Property<URI> = objects.property()
-
-    @get:Optional
-    @get:Input
-    val username: Property<String> = objects.property()
-
-    @get:Optional
-    @get:Input
-    val password: Property<String> = objects.property()
-
-    @get:Optional
-    @get:Input
-    val packageGroup: Property<String> = objects.property()
-
-    @get:Optional
-    @get:Input
-    val stagingProfileId: Property<String> = objects.property()
-
-    @get:Input
-    val repositoryName: Property<String> = objects.property()
-
-    @get:Internal
-    val clientTimeout: Property<Duration> = objects.property()
+constructor(objects: ObjectFactory, extension: NexusPublishExtension, repository: NexusRepository, private val serverUrlToStagingRepoUrl: MutableMap<URI, URI>) :
+        BaseOperationOnNexusStagingRepository(objects, extension, repository) {
 
-    @get:Internal
-    val connectTimeout: Property<Duration> = objects.property()
-
-    init {
-        serverUrl.set(repository.nexusUrl)
-        username.set(repository.username)
-        password.set(repository.password)
-        packageGroup.set(extension.packageGroup)
-        stagingProfileId.set(repository.stagingProfileId)
-        repositoryName.set(repository.name)
-        clientTimeout.set(extension.clientTimeout)
-        connectTimeout.set(extension.connectTimeout)
-        this.onlyIf { extension.useStaging.getOrElse(false) }
-    }
+    private val stagingRepositoryId: Property<String> = repository.stagingRepositoryId
```

Comment:
I'm not sure about the best way to bind stagingRepositoryId in `InitializeNexusStagingRepository` and `Repository.stagingRepositoryId`, to be able to set value in task to set it in Extension/Repository. I initially just set `stagingRepositoryId: Property<String> = repository.stagingRepositoryId`, but I don't like sharing a reference, so maybe you have any better idea.

Reply:
` = objects.property().convention(repository.stagingRepositoryId)` ?

Reply:
Hmm, Unless I don't understand `convention()` properly, I will not be able to propagate value back to repository once it is set in the task. I need the other way round connection: task -> repository.

Reply:
Oh, now I understand your intention.

Can you please clarify how do you mean the task should be used?
I think the user does not know `stagingRepositoryId` before the task is executed, so I don't see how the user would configure the property.

Reply:
User could override the property in extension.repository to for example close or releate a requested staging repository with id known from the previous (failed) closing operation.

Reply:
What if I want to "discard old repository", "initialize new one", and "close it" in one go?

Use case: "produce next release candidate"
If old release candidate exists, then it might make sense to drop the previous candidate. Of course, the new candidate should be pushed to a different repository => need two different repository ids for the same nexus server within a same build.

Use case: "publish release and drop all release candidate repositories".
If release candidates are kept till the release is promoted (e.g. to enable comparison between RC artifacts), then "promote release" tasks needs to be able to drop multiple distinct repositories => again it is a case for "more than one staging id per nexus server within a single Gradle invocation".

Reply:
How problematic would it be to do the clean up in the second call? E.g.
> gw release* -PyourID* && gw dropAllRepositories*
(assuming there will be that task one day)

While technically it is possible to call close/release/drop commands in Nexus with multiple repository IDs it's not a case with simple init/publish/close/release use case. On the other hand I can't find any sensible obstacles to use `ListProperty<String>` in the extension and have init to set just one element (there could be also some helper method to set just one).

Reply:
> How problematic would it be to do the clean up in the second call? 

Users are not always comfortable with storing passwords/keys, so "release / drop" might require the user to type the password (or provide it with environment/property).
So your example with `&&` would be much more complicated, as there are multiple credentials involved. In Apache case the credentials are for `dist.apache.org` (SVN), `repository.apache.org` (Nexus), `gitbox.apache.org` (Git) all of which participate in the release candidate/release publish.

Having a single command helps there.

Then there's "error recovery". If the actions are split into multiple commands, it makes it more complicated to "resume after failure". It looks like the user would have to analyze which command failed in order to retry or rollback.

Reply:
Why do we need this property in this class at all? Wouldn't it suffice to set it in `NexusRepository`? In addition, I think we should write it to a file and read it from there in `NexusRepository` by default. That way, one could initialize and publish in the first Gradle invocation, do some manual testing, and close and release without passing the `stagingRepositoryId` manually. Thought?

Reply:
Currently `NexusRepository` is only passed to the constructor to collect some required values. The reference is not kept. However, if you are fine with that I will refactor the code.

Regarding keeping it in the file I propose to postpone it to Milestone 2 - it's not MVP.

Reply:
Changed.